Created `Current-guide.md` (#86)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/Current-guide.md

# `Current` Guide - how to use it in Babylon iOS Project

> ### Where is that? 

Reviewer: What is `that`? Could this be rephrased?

Author: `that` is `Current`, the subject of the previous line, which I prefer to avoid to repeat.

Reviewer: ```suggestion
### Where can I find it?
```

Reviewer: Also this section feels more like an overview that just a where to find the file. Maybe something like 

```suggestion
### Overview
```

Author: I accepted this but I don't agree using first or second person. 
I've checked and I don't see any of the Cookbook using this way to explain. (expect for the Olivier's about framework).

- `Current` is the global instance of the `World` type which is located in the `BabylonDependencies.framework`. 

> Its purpose is to provide convenient access to the instances that should be shared between different parts of the app (cross-cutting concerns). For more information please read the [How Control the World](/Cookbook/Proposals/ControlTheWorld.md) document.

Reviewer: I was re reading this and I think we could improve this by saying what goes into Current and what the client can expect from it. Right now there is a link to the proposal, but the proposal might be outdated and not up to our current standards. So if this is the document that explains how to use Current it should also provide a tiny window into its contents.

Author: 1.
> what goes into Current

Isn't this part clear? 
> convenient access to the instances that should be shared between different parts of the app

2.
> what the client can expect from it
I don't understand it. 
Again:
> ...access to the instances that should be shared...

3.
> might be outdated and 
the part that is outdated has been pointed by @inamiy  in [this comment](https://github.com/Babylonpartners/ios-playbook/pull/86#pullrequestreview-243068580) . I'm thinking how to cover that.

> not up to our current standards

I don't understand it. 

4. 
> it should also provide a tiny window into its contents
IMO pointing where to find is already enough to the Engineer find it's contents. But considering the tiny window, there is one already:
```swift 
let locale = Locale(identifier: "haw_US")
Current = World(autoupdatingLocale: locale)
```
I'm assuming (by the sequence of the Cookbook) that who's reading it, has already read the Proposal already mentioned. So no need to repeat.

## `Do`s and `Don't`s

> #### Dos

Reviewer 2: ```suggestion
#### Do
```

> - access `Current` **only** from topmost `AppDelegate` until `Builder`s layer

Reviewer: ```suggestion
- You should only access `Current` in builders, not in ViewModels or Renderers. The AppDelegate is included
```

Author: - lowercase, and not in first person
- the **not in** is part of the `Don't` section

> - pass the required dependencies down the chain;

Reviewer: ```suggestion
- Pass the required dependencies down the chain. The builder accesses Current and it injects the specific dependencies down the chain
```

Author: It's repeating "down the chain" twice.

Author: also as items I prefer to have them as 
- lowercase

Reviewer: Items in a list are usually uppercase https://www.quickanddirtytips.com/education/grammar/formatting-vertical-lists

Reviewer 3: List items AFAIK are uppercase. 🤔

Author: @diegopetrucci from the link you provided I see a capitalised list 👍 :

<img width="228" alt="Screen Shot 2019-06-11 at 14 50 59" src="https://user-images.githubusercontent.com/17405823/59277576-5c983380-8c58-11e9-9e4e-2b65ddedb3e1.png">

But giving a quick read... I clicked the link Chicago Manual Style... https://www.quickanddirtytips.com/education/grammar/grammar-style-issues

> One book discusses capitalization, another partially discusses punctuation, and so on. Only the [Chicago Manual of Style](https://www.quickanddirtytips.com/education/grammar/grammar-style-issues) gives the topic the respect it deserves. 

... I see a lowercased list 

<img width="354" alt="Screen Shot 2019-06-11 at 14 52 53" src="https://user-images.githubusercontent.com/17405823/59277780-ac76fa80-8c58-11e9-9ddc-6af588ea2838.png">

I change it as long it adds value to the document and improve the understanding.

> - create a customised version to fit your specific needs **for tests**;

Reviewer 4: ```suggestion
- create a customized version to fit your specific needs **for tests** and set it in the test `setUp` method
```

> - create a customized version to fit your specific needs **for tests** and set it in the test `setUp` method

Reviewer: ```suggestion
- In your tests, you are free to create a customized version of Current that fits your specific needs. Set it up in the `setUp` method. After you're done, reset it in `tearDown`. Eg
```

```swift
let locale = Locale(identifier: "haw_US")

> Current = World(autoupdatingLocale: locale)

Reviewer 5: I'm not sure about this one though. Mutating the global `Current` would cause issues in tests if running in parallel - we've seen this issue with snapshots when changing the design library theme in Current and I don't see why the same wouldn't happen when doing it this way

Reviewer 5: (If this is correct, it should be added to the don'ts list)

Author: If it can’t be mutated then the 
```#if DEBUG
var Current```
doesn’t make sense anymore. 
It has to be fixed or a `restore` mechanism put in place..
Good point tho 👍

Reviewer 5: Hmmm, you're right. If we make it immutable, then we're essentially back where we were but with normal DI, but without the long initialisers.

Author: @Ben-Henshall I think @ilyapuchka 's suggestion here addresses your concerns about the `Current` mutations..
```- reset the instance of `Current` to the default in the `tearDown` method```

Given that, I think we can resolve this now right?

Reviewer 4: Yes, I think it can be one of the reasons we can have instability in tests. I'm more convinced in that than that it is mutating `Current` in the first place - tests that are running in parallel should be running in separate processes thus not sharing a memory, and tests running in one process should be running sequentially, otherwise, I don't know how parallelization works in general 🤷‍♂

```

> 

Reviewer 4: ```suggestion
- reset the instance of `Current` to the default in the `tearDown` method
```

> - reset the instance of `Current` to the default in the `tearDown` method

Reviewer: ```suggestion
```

Author: it shouldn't be removed. I prefer the clarity of the separation **do** and **don't**

Reviewer: My problem here is that it still feels part of the previous step, as we are talking about how to use current in tests.

Author: You're correct. It is indeed part of the previous step. 
This commit (https://github.com/Babylonpartners/ios-playbook/pull/86/commits/27b1f6b10e55e9504ba3d68bb3ae95912dda7b56) I addressed that specific `setUp` and `tearDown` methods changes. 
Thanks.

#### Don't

> - access `Current` from any other layer other than `Builder`. All the other layers (`ViewModel`, `FlowController`, `ViewController`, `Model`, `BusinessController`) continue to receive their dependencies through injection when created by the builder, preferably at initialisation time;

Reviewer 5: What about decodable/encodable objects? Was there a consensus on this?

Author: The consensus is: just use it on Builders. The first *don’t* mentions `Model` ... should receive its dependencies through injection.

Reviewer 6: @AliSoftware Reminder to self: make a Danger rule for this.

> - inject `current` as a dependency;

Reviewer 5: ```suggestion
- inject `Current` as a dependency. Individual parts of `Current` should be passed instead.
```

Author: Isn’t it repeating the previous? 
> .. continue to receive their dependencies through injection when created by the builder, preferably at initialisation time;

Reviewer 5: Yep!

> - inject `current` as a dependency;

Reviewer: ```suggestion
```

Author: this is very important. It shouldn't be removed.

Reviewer: But it’s already stated in the previous point

Author: it shouldn't be removed. I prefer the clarity of the separation do and don't

> - try to mutate `Current`. In `Release` configuration it's a constant `let`;

Reviewer: ```suggestion
- try to mutate `Current` (excluding Tests). In `Release` configuration it's a constant `let`;
```

# `Current` Guide - how to use it in Babylon iOS Project

### Where can I find it?
- `Current` is the global instance of the `World` type which is located in the `BabylonDependencies.framework`. 
Its purpose is to provide convenient access to the instances that should be shared between different parts of the app (cross-cutting concerns). For more information please read the [How Control the World](/Cookbook/Proposals/ControlTheWorld.md) document.

## `Do`s and `Don't`s

#### Do
- access `Current` **only** from topmost `AppDelegate` until `Builder`s layer
- pass the required dependencies down the chain;

> - create a customized version to fit your specific needs **for tests** and set it in the test `setUp` method. Please reset it on the `tearDown` method 

Reviewer 7: I think this is no longer the right approach since we have started to run unit tests in parallel. Am I right?

Author: We have parallel **and** serial tests running. Imagine the tests `a`, `b`, and `c` in a process and `d`, `e` and `f` in another.. 

When `a` finishes, it's important to reset `Current` in `tearDown`. 
@TheAdamBorek ^
@AliSoftware 
@ilyapuchka

Reviewer 6: I confirm what @daniloaliberti said, that's also my understanding of how they are distributed across sims. Parallel testing doesn't mean you shouldn't reset the state.

But that's not the reality for all of them. These `async` properties have their initial value provided in the `SharedAppDelegate` through the method `setupCurrent`. Only after that, the whole world is ready to be used.

## `Do`s and `Don't`s

#### Do
- access `Current` **only** from topmost `AppDelegate` until `Builder`s layer
- pass the required dependencies down the chain;
- create a customized version to fit your specific needs **for tests** and set it in the test `setUp` method. Please reset it on the `tearDown` method 
```swift
override func setUp() {
    let utc = TimeZone(identifier: "UTC")!
    let locale = Locale(identifier: "en_GB")
    Current = World(autoupdatingTimeZone: utc, autoupdatingLocale: locale)
}

override func tearDown() {
    Current = .production // which returns a new `World()`
}
```
#### Don't
- access `Current` from any other layer other than `Builder`. All the other layers (`ViewModel`, `FlowController`, `ViewController`, `Model`, `BusinessController`) continue to receive their dependencies through injection when created by the builder, preferably at initialisation time;
- inject `current` as a dependency;

> - try to mutate `Current` (excluding Tests). In `Release` configuration it's a constant `let`;

Reviewer 6: Should we remove the `Current.changeTheme` method from `World` then?

Is there a difference in that sentence between "mutating the content of `Current`" and "replace the `Current` instance with another one"? iow, are we both prohibited from doing `Current = someNewValue` but also `Current.changeWhatever` – assuming `changeWhatever` mutates some properties of the `Current` instance?

I think this point deserves clarification in this sentence to disambiguate between the two cases.

Author: Perfect point @AliSoftware 

There is a bunch of `set` methods:
`setBiometricAuth`, `setNetworkDependencies`, `setEnvironment`, `setAppConfiguration` and also as you said the `changeTheme`. 

These methods change **values** inside Current's private mutable properties that are consumed by its public immutable computed ones. Basically value semantics over mutability.

Answering your question: both are prohibited. The only entity allowed to run these `set` methods is the SharedAppDelegate.

I wonder if changing the:
> (excluding Tests)

to 

> (excluding Test targets and `SharedAppDelegate.setupCurrent` methods)

Would it disambiguate that point?

> - try to mutate `Current` (excluding Test targets and `SharedAppDelegate.setupCurrent` methods). In `Release` configuration it's a constant `let`;

Reviewer 6: I think it would be useful to specify `excluding test targets and setupCurrent` indeed, but I don't think that would be enough to clarify that both "Replacing the Current instance altogether" and "mutating the content of Current" are prohibited elsewhere.

I mean if you look at the code you can eventually see that `World` is a `struct` so has value semantics and thus when it's `let` (in Release) you won't be able to mutate even its content, so you can _guess_ and _conclude_ that indeed both should not be allowed anyway, but making it explicit in such a document dedicated to `Current` still feels useful to me.

```suggestion
- Mutate `Current`'s properties – or re-affect it to another instance – after `SharedAppDelegate.setupCurrent()`. You are only allowed to mutate `Current` in tests, not in app. In `Release` configuration it's a constant `let` anyway to prevent mutation;
```
